src/OAuth/AuthorizationMiddleware.php

```diff
@@ -0,0 +1,191 @@
+
+    /**
+     * @param callable                   $nextHandler
+     * @param CredentialStorageInterface $credentialStorage
+     * @param ClientInterface            $httpClient
+     * @param string                     $clientId
+     * @param string                     $clientSecret
+     */
+    public function __construct(
+        callable $nextHandler,
+        CredentialStorageInterface $credentialStorage,
+        ClientInterface $httpClient,
+        string $clientId,
+        string $clientSecret
+    ) {
+        $this->nextHandler       = $nextHandler;
+        $this->credentialStorage = $credentialStorage;
+        $this->httpClient        = $httpClient;
+        $this->clientId          = $clientId;
+        $this->clientSecret      = $clientSecret;
+    }
+
+    /**
+     * Creates a callable that wraps the middleware with the next handler of @see HandlerStack
+     *
+     * @param CredentialStorageInterface $credentialStorage
+     * @param ClientInterface            $httpClient
+     * @param string                     $clientId
+     * @param string                     $clientSecret
+     *
+     * @return callable
+     */
+    public static function wrapped(
+        CredentialStorageInterface $credentialStorage,
+        ClientInterface $httpClient,
+        string $clientId,
+        string $clientSecret
+    ): callable {
+        return function (callable $nextHandler) use ($credentialStorage, $httpClient, $clientId, $clientSecret) {
+            return new self($nextHandler, $credentialStorage, $httpClient, $clientId, $clientSecret);
+        };
+    }
+
+    /**
+     * @param CommandInterface $command
+     *
+     * @return PromiseInterface
+     */
+    public function __invoke(CommandInterface $command): PromiseInterface
+    {
+        if (empty($command['referenceID'])) {
+            return new RejectedPromise(UnauthorizedException::missingReferenceId());
+        }
+
+        $referenceID = $command['referenceID'];
+        $credential  = $this->credentialStorage->get($referenceID);
+
+        if (null === $credential) {
+            return new RejectedPromise(UnauthorizedException::missingCredential($referenceID));
+        }
+
+        $command = $this->authorizeCommand($command, $credential);
+
+        /** @var PromiseInterface $promise */
+        $promise = ($this->nextHandler)($command);
+
+        // If authentication fails, we refresh the token and try again
+        return $promise->otherwise(function (CommandException $exception) use ($command, $credential) {
+            $response = $exception->getResponse();
+
+            // Forward non 401 errors
+            if (null === $response || 401 !== $response->getStatusCode()) {
+                throw $exception;
+            }
+
+            $credential = $this->refreshToken($credential);
+            $command    = $this->authorizeCommand($command, $credential);
+
+            // Try again
+            return ($this->nextHandler)($command);
```

Comment:
Nope I don't think this will happen a lot. We'll try to limit the number of instances to 1.